### PR TITLE
Add additionalMounts for GitHub Action Output Mount

### LINKED ISSUE #180: 'Save-State' Command is Deprecated
The 'save-state' command is deprecated: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

I have had to update other actions due to this deprecation as well, if you would like me to update these commands please let me know :).

### PR DESCRIPTION
**Pull request by stephenandary**
@stuartleeks 

Beginning PR to add additional mounts required for GitHub Output in CI. 

Recreating <https://github.com/devcontainers/ci/pull/187> work with a non-stale base. 

WIP(DevContainerCliUpArgs - initial commit to add additional mounts for GitHub Action output (azdo-task/scripts/build-package.sh#L49->L64

There are a few deletions I did not make: 
  - Lines 188, 195, and 200 in common/src/dev-container-cli.ts.
  - Lines 177 to 125 were formatted on save. 
  
---
@stuartleeks update:
Fixes #180, #188, #195

### CODE REVIEW

**Comment by stephenandary** [comment]
I am unsure how to go about testing this, some direction here would be wonderful. Thank you!

File: common/src/dev-container-cli.ts
Comment (stuartleeks) [inline common/src/dev-container-cli.ts:195]: Is this an intentional removal?

Comment (stephenandary) [reply common/src/dev-container-cli.ts:195]: Lines 188, 195, and 200 in common/src/dev-container-cli.ts, I did not remove and this was from a clean rebase so I thought I should push first with them removed to ask. Should I add these lines back? The language server was indicating they were unused to me.

**Comment by stuartleeks** [comment]
In terms of testing, this is a good nudge to add something to the docs - thank you!

I just ran a test and discovered that there are a couple of things that have snuck through that prevent running tests in a fork and have created #220.

Once that is merged, you can go to the Actions tab for your fork and you will see the various workflows listed:

image

Click on `CI (branch)` as shown in the image above, and then click on the `Run workflow` button:

image

This will pop up some options. Select your PR branch, enter a number in the PR number box, and untick the `run full tests` box (the last bit is important):

image

This should do a run for most of the GitHub action tests. The Azure DevOps tests require an environment to be set up and configured which can be triggered by maintainers against a PR.

**Comment by stephenandary** [comment]
@stuartleeks should I add back those deleted lines? I am not sure if it was my ide, or if a merge happened within the time I had synced and pushed.

**Comment by stuartleeks** [comment]
@stephenandary - yes, those lines are needed

**Comment by stephenandary** [comment]
@stuartleeks in my tests I am getting hit by some permission and 429 errors, can't get past the first block. 
image

**Comment by stuartleeks** [comment]
@stephenandary - https://github.com/devcontainers/ci/pull/220 has been merged now, so if you merge that into your PR branch you should be able to use the steps above to run the subset of GH tests that don't need credentials on your fork (i.e. without needing someone else to trigger a test run)

In terms of tests, yes I think it would be good to add a test showing that GITHUB_OUTPUT is working from a `runCmd`

**Comment by stephenandary** [comment]
@stuartleeks do you recall what the mount path should be from <https://github.com/devcontainers/ci/pull/187#issuecomment-1356102335>?

File: azdo-task/vss-extension.json
Comment (stephenandary) [inline azdo-task/vss-extension.json:5]: Is this correct?

Comment (stuartleeks) [reply azdo-task/vss-extension.json:5]: `0.0.0` is the latest value on `main` (https://github.com/devcontainers/ci/blob/main/azdo-task/vss-extension.json#L5) from the PR to enable building on forks.

I'm confused about why this is showing as a change in your PR though

File: common/src/dev-container-cli.ts
Comment (stuartleeks) [reply common/src/dev-container-cli.ts]: Strange - they are present on latest `main` code: https://github.com/devcontainers/ci/blob/main/common/src/dev-container-cli.ts#L195

Comment (stephenandary) [reply common/src/dev-container-cli.ts]: Indeed

File: azdo-task/vss-extension.json
Comment (stephenandary) [reply azdo-task/vss-extension.json:5]: There must have been a merge right after I synced. May just be bad timing?

**Comment by stephenandary** [comment]
> @stuartleeks do you recall what the mount path should be from [#187 (comment)](https://github.com/devcontainers/ci/pull/187#issuecomment-1356102335)?

@stuartleeks, just bubbling this up again, as it will save me time having to figure this out if you know. 

I also merged main, I was 1 commit behind which explains those little changes that were required.

**Comment by stuartleeks** [comment]
There was some discussion around the `GITHUB_OUTPUT` handling in https://github.com/devcontainers/ci/issues/188 - is that what you mean?

**Comment by stephenandary** [comment]
@stuartleeks based on the approach you highlighted in <https://github.com/devcontainers/ci/issues/188>, I am unsure what tests are appropriate to modify. 

I have added mounts in `common/__tests__`, (this does not test `/mnt/github/output`). I noticed that there are also `github-tests`. Should I be adding a mount or environment variable into the `devcontainer.json`s for the various scenarios there?  By the description of the solution approach, it seems this should be automatic; I am not understanding how this is currently automatic and how only the common test will cover it.

**Comment by stuartleeks** [comment]
The tests in `common\__tests__` are unit tests, whereas the contents under `github-tests` are used by the integration tests in CI. 
The folder is slightly misnamed now as the content is used by both GitHub workflow tests and Azure Devops pipeline tests! The GitHub workflow to exercise them is in `.github/workflows/ci_common.yml` and the Azure DevOps pipeline to use them is in `.azure-devops/azure-pipelines.yml`

**Comment by stephenandary** [comment]
Thanks @stuartleeks! I think where I am getting confused a bit is to whether a user will need to add the GITHUB_OUTPUT mount in the devcontainer.json to use env variable, or if there is a way that does not require user input. I am going to go down the road of adding the mount this morning, and hopefully, this becomes more clear during that exercise.

**Comment by stephenandary** [comment]
@stuartleeks ,
To highlight where I am currently at, your proposed approach from #188 

- mount the host's GITHUB_OUTPUT location to /mnt/github/output inside the dev container
- set the GITHUB_OUTPUT environment variable in the container to /mnt/github/output

Currently, in integration tests, I am thinking of adding a mount along the lines of: 
```json
"mounts": [
		// Keep command history 
		"source=build-args-bashhistory,target=/home/vscode/commandhistory",
		// Mount to utilize GITHUB_OUTPUTS
		"source=/home/runner/work/${REPO}/${REPO}/.github/${GITHUB_ACTION}/${STEP_NAME}/output, target=/mnt/github/output"
],
```
I am not sure if this is the correct approach so trying to document in PR as I go. I imagine I would need to update the DOCKERFILE to reflect the GITHUB_OUTPUT env variable. 

Does this make sense to you?

**Comment by stuartleeks** [comment]
I imagined that the action code would determine the values for `GITHUB_OUTPUT` etc on the host and automatically add that as an additional mount when invoking the CLI, at the same time also setting the `GITHUB_OUTPUT` environment variable for the container to point to the mounted location. And then to do this for `GITHUB_STATE` etc

**Comment by stephenandary** [comment]
@stuartleeks, to make sure I understand, I should not worry about runArgs and buildArgs tests, and just update the `.github/workflows/ci_common.yml` and  `.azure-devops/azure-pipelines.yml` actions?

**Comment by stuartleeks** [comment]
I'd add a new test job in the GH workflow (`ci_common.yml`) for testing that `GITHUB_OUTPUT` etc are working correctly. In the `runCmd` for that job, set a variable via `GITHUB_OUTPUT` and then check in a subsequent step that the output has been set

File: .github/workflows/ci_common.yml
Comment (stephenandary) [inline .github/workflows/ci_common.yml:1091]: Could have used `run: echo "test=hello" >> "$GITHUB_OUTPUT"` instead, but kept with runArg per other cases. 'uses: ./` may be incorrect

**Comment by stephenandary** [comment]
@stuartleeks my tests ran though I sill get a 
```
`Warning: The `save-state` command is deprecated and will be disabled soon. 
Please upgrade to using Environment Files. For more information see: 
https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```
during the Build In Devcontainer job.

File: .github/workflows/ci_common.yml
Comment (stephenandary) [reply .github/workflows/ci_common.yml:1105]: Yes, I will do this, I also forgot to add the new tests into the test list. I already fixed that

File: .github/workflows/ci_common.yml
Comment (stuartleeks) [reply .github/workflows/ci_common.yml:1092]: `uses: ./` tells the GH workflow to run the step using the devcontainer/ci action. We reference it by path here so that we pick up the code from the PR branch rather than a published version of the action

**Comment by stuartleeks** [comment]
@stephenandary I think the `save-state` warning is coming from here. I've created #221 to bump the version of @actions/core which should cause it to use `GITHUB_STATE` rather than `save-state`.

Once the `GITHUB_OUTPUT` mounting/env vars are being set in the dev container, the `set-output` lines [in the CI script](https://github.com/devcontainers/ci/blob/4eceb21758dae024fe97bece28edaf76c19a2e18/azdo-task/scripts/build-package.sh#L57-L59) can be converted to use `GITHUB_OUTPUT` instead, and that should remove the `set-output` warnings

**Comment by stephenandary** [comment]
> @stephenandary I think the `save-state` warning is coming from here. I've created #221 to bump the version of @actions/core which should cause it to use `GITHUB_STATE` rather than `save-state`.
> 
> Once the `GITHUB_OUTPUT` mounting/env vars are being set in the dev container, the `set-output` lines [in the CI script](https://github.com/devcontainers/ci/blob/4eceb21758dae024fe97bece28edaf76c19a2e18/azdo-task/scripts/build-package.sh#L57-L59) can be converted to use `GITHUB_OUTPUT` instead, and that should remove the `set-output` warnings

@stuartleeks I checked this yesterday and though actions/core was a version greater than req for GITHUB_OUTPUT already, I must have been mistaken. 

Versions are not currently being set in tests, I will get back to it this morning in a bit.

File: azdo-task/scripts/build-package.sh
Comment (stephenandary) [inline azdo-task/scripts/build-package.sh:54]: @stuartleeks these are the lines where the version is not grabbed which causes build job to fail, they should not be failing due to anything I changed because this is unrelated code. On occasion it runs, and on occasion it does not.

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh:54]: My mistake, I think it may actually be the lines I added for changing output method:

```sh
  [2023-03-03T12:20:46.703Z] --set-patch-version not set
  
  [2023-03-03T12:20:46.717Z] VERSION_MAJOR=0
  VERSION_MINOR=3
  
  [2023-03-03T12:20:46.717Z] 
  VERSION_PATCH=0
  version=0.3.0 >> 
  version_short=0.3 >> 
  version_major=0 >> 
  ```
  
  I thought the required version was 1.1.0 or greater for actions/core, but maybe I missread and .10 is needed for this as you have set in #221

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh:54]: I did misread, I have updated quotes to be appropriate for command, and will wait for #221 to be merged to test again.

File: azdo-task/scripts/build-package.sh
Comment (stephenandary) [inline azdo-task/scripts/build-package.sh:57]: @stuartleeks test is failing here and I am not sure how to progress: 

Error:
```sh
  \ \   / /__ _ __ ___(_) ___  _ __  
   \ \ / / _ \ '__/ __| |/ _ \| '_ \ 
    \ V /  __/ |  \__ \ | (_) | | | |
     \_/ \___|_|  |___/_|\___/|_| |_|
                                     
  [2023-03-03T13:27:44.417Z] --set-patch-version not set
  [2023-03-03T13:27:44.417Z] 
  [2023-03-03T13:27:44.431Z] VERSION_MAJOR=0
  VERSION_MINOR=3
  VERSION_PATCH=0
  [2023-03-03T13:27:44.431Z] ./scripts/build-package.sh: line 57: $GITHUB_OUTPUT: ambiguous redirect
  [2023-03-03T13:27:44.431Z] 
  Error: Dev container exec: Command failed: bash -c echo "Starting"
  ./scripts/build-test-package.sh (exit code: undefined)
  An error occurred running a command in the container.
  Error: Command failed: bash -c echo "Starting"
  ./scripts/build-test-package.sh
  ```

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh:57]: trying : ` $env:GITHUB_OUTPUT`

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh:57]: Starting here: <https://github.com/community/community/discussions/36746>

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh:57]: My runner is > 2.297.0

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh]: quoting $GITHUB_OUTPUT in build-package.sh results in this error:
```sh
 \ \   / /__ _ __ ___(_) ___  _ __  
   \ \ / / _ \ '__/ __| |/ _ \| '_ \ 
    \ V /  __/ |  \__ \ | (_) | | | |
     \_/ \___|_|  |___/_|\___/|_| |_|
                                     
  [2023-03-03T13:54:26.162Z] --set-patch-version not set
  [2023-03-03T13:54:26.162Z] 
  [2023-03-03T13:54:26.180Z] VERSION_MAJOR=0
  VERSION_MINOR=3
  VERSION_PATCH=0
  ./scripts/build-package.sh: line 57: : No such file or directory
  [2023-03-03T13:54:26.181Z] 
  Error: Dev container exec: Command failed: bash -c echo "Starting"
  ./scripts/build-test-package.sh (exit code: undefined)
  An error occurred running a command in the container.
  Error: Command failed: bash -c echo "Starting"
  ./scripts/build-test-package.sh
  ```
  
  I don't have any more time this morning to continue with this, I will do my best to make some over weekend, or early next week.

Comment (stuartleeks) [reply azdo-task/scripts/build-package.sh]: I think that there are probably a few small tweaks to your PR to get this working. Would you be open to me pushing some commits to your PR branch?

Comment (stephenandary) [reply azdo-task/scripts/build-package.sh]: 💯, thank you!

**Comment by stuartleeks** [comment]
As discussed in one of the earlier comments, I pushed a couple of commits to add the mounts/env vars for `GITHUB_OUTPUT` and that resolved the errors you were seeing around quoting `$GITHUB_OUTPUT` in the script.

I also pushed the changes from #221 here, so I've closed that PR in favour of this one. I pushed this to the `pr-219` branch in the upstream repo to kick of the tests and found a few more action upgrades that were needed to remove some more warnings around NodeJS versions, `set-state` etc. I've pushed those and [this run](https://github.com/devcontainers/ci/actions/runs/4330592051) is a clean run 🎉 

Thanks for all your work on this!

**Comment by stephenandary** [comment]
@stuartleeks this is so wonderful! Thank you for helping me learn. It would have taken me a while to figure out those additional changes and I'm grateful you could take this over the finish line. It was an enjoyable experience to work on, and thats all you! P.S. this was my first Open Source PR ever haha, going to save this one as a memento!

**Comment by stuartleeks** [comment]
@stephenandary - you are welcome. Thanks again for your work on this, and here's to many more OSS PRs 😁

**Review by chrmarti — APPROVED** [review]
LGTM!

**Pull request merged**
